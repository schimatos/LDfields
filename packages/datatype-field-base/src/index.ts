import { LDfieldBase } from '@ldfields/field-base';

/**
 * LDField input specifically for datatype inputs
 */
export abstract class DatatypeInput<
  Rendered,
  Props extends {
    termType: string;
    datatype: string;
    value: string;
    [key: string]: string;
  },
  ExtraData = never
> extends LDfieldBase<Rendered, Props, ExtraData> {
  /**
   * The datatype(s) that the field supports.
   */
  abstract supportedDatatypes: Record<string, boolean>;

  /**
   * The field entries that the Datatype input
   * is capable of modifying.
   */
  modifies: (keyof Props & string)[] = ['value'];

  supports(props: Props) {
    return props.termType === 'Literal' && this.supportedDatatypes[props.datatype];
  }
}
