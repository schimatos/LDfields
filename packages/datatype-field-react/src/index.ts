export class GenericDatatype extends DatatypeInput {
  priority = 50;

  supportedDatatypes = {
    [xsd.integer]: true,
    [xsd.float]: true,
    [xsd.decimal]: true,
    [xsd.dateTime]: true,
  };

  Field({ props, constraints, onChange }) {
    // TODO: Bring in min count, min length etc.
    return (
      <input
        type={props.datatype === xsd.dateTime ? "dateTime" : "number"}
        value={props.value}
        onChange={(e) => {
          onChange({ value: e.target.value });
        }}
      />
    );
  }