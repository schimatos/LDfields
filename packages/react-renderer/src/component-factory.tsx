import React from 'react';
import type { LDfieldBase } from '@ldfields/field-base';
import type { FieldProps } from '@ldfields/field-base/types';

/**
 * Renders the set of components based on the delegated fields.
 * @param components The set of LDfield bases.
 */
export function ComponentFactory<
  Props extends { [key: string]: string | undefined; },
  ExtraData = never
>(
  components: LDfieldBase<JSX.Element, Props, ExtraData>[],
) {
  return ({ onChange, ...props }: FieldProps<Props, ExtraData>) => (
    <>
      {
        components.map((Field: LDfieldBase<JSX.Element, Props, ExtraData>, i: number) => (
            <Field.Field {...props} key={i} onChange={(p) => onChange({ ...props.props, ...p })} />
        ))
      }
    </>
  );
}
