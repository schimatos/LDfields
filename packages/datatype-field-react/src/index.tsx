import React from 'react';
import type { FieldProps } from '@ldfields/field-base/types';
import { DatatypeInput } from '@ldfields/datatype-field-base';
// TODO: Use typed on2ts values here
import xsd from '@ontologies/xsd';

// TODO: Finish up this
export class GenericDatatypeField<
  Props extends { [key: string]: string; termType: string; datatype: string; value: string; },
  ExtraData = never,
> extends DatatypeInput<JSX.Element, Props, ExtraData> {
  priority = 50;

  supportedDatatypes = {
    [xsd.integer.value]: true,
    [xsd.float.value]: true,
    [xsd.decimal.value]: true,
    [xsd.dateTime.value]: true,
  };

  // eslint-disable-next-line class-methods-use-this
  Field({ props, onChange }: FieldProps<Props>) {
    // TODO: Bring in min count, min length etc.
    return (
      <input
        type={props.datatype === xsd.dateTime.value ? 'dateTime' : 'number'}
        value={props.value}
        onChange={(e) => {
          onChange({ value: e.target.value });
        }}
      />
    );
  }
}
