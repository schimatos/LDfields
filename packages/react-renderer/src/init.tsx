import type { InputDelegator } from '@ldfields/delegator';
import type { Constraints } from '@ldfields/field-base/types';
import type { GetComponentOut, InitParams, RendererState } from '../types';
import { ComponentFactory } from './component-factory';

class ReactInputDelegator<
  Props extends { [key: string]: string; },
  ExtraData = never
> {
  GetComponent: (...[props, constraints, data]: InitParams<Props, ExtraData>) => GetComponentOut<
    Props,
    ExtraData
  >;

  private delegator;

  private Component = ComponentFactory<Props, ExtraData>([]);

  constructor(
    InputDelegator: new () => InputDelegator<
      JSX.Element,
      Props,
      ExtraData,
      Constraints<Props>
    >,
  ) {
    this.delegator = new InputDelegator().delegate;

    this.GetComponent = (...props: InitParams<Props, ExtraData>):
      GetComponentOut<Props, ExtraData> => {
      const {
        modifiable, required, delegation, delegationChange,
      } = this.delegator(...props);
      return {
        modifiable,
        required,
        delegationChange,
        Component: delegationChange ? ComponentFactory(delegation) : this.Component,
      };
    };
  }
}

/**
 * Factory to generate the init function for a reducer
 * @param InputDelegator LDfield input delegator.
 */
export function initFactory<
  Props extends { [key: string]: string; },
  ExtraData = never
>(
  InputDelegator: new () => InputDelegator<
    JSX.Element,
    Props,
    ExtraData,
    Constraints<Props>
  >,
) {
  const { GetComponent } = new ReactInputDelegator(InputDelegator);
  // TODO [Future]: Add constraint type parameters
  return function init([props, constraints, data]: InitParams<Props, ExtraData>):
    RendererState<Props, ExtraData> {
    return {
      GetComponent,
      ...GetComponent(props, constraints, data),
      cache: [props],
      props,
    };
  };
}
