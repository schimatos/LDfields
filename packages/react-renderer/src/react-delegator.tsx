import type { InputDelegator } from '@ldfields/delegator';
import type { Constraints } from '@ldfields/field-base/types';
import type { GetComponentOut, InitParams } from '../types';
import { ComponentFactory } from './component-factory';

export class ReactInputDelegator<
  Props extends { [key: string]: string | undefined },
  ExtraData = never
> {
  GetComponent: (
    ...[props, constraints, data]: InitParams<Props, ExtraData>
  ) => GetComponentOut<Props, ExtraData>;

  private delegator;

  private Component = ComponentFactory<Props, ExtraData>([]);

  constructor(
    InputDelegator: new () => InputDelegator<
      JSX.Element,
      Props,
      ExtraData,
      Constraints<Props>
    >,
  ) {
    this.delegator = new InputDelegator().delegate;

    this.GetComponent = (
      ...props: InitParams<Props, ExtraData>
    ): GetComponentOut<Props, ExtraData> => {
      const {
        modifiable,
        required,
        delegation,
        delegationChange,
      } = this.delegator(...props);
      return {
        modifiable,
        required,
        delegationChange,
        Component: delegationChange
          ? ComponentFactory(delegation)
          : this.Component,
      };
    };
  }
}
