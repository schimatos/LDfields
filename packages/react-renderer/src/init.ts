import type { InputDelegator } from '@ldfields/delegator';
import type { Constraints } from '@ldfields/field-base/types';
import type { InitParams, RendererState } from '../types';
import { ReactInputDelegator } from './react-delegator';

/**
 * Factory to generate the init function for a reducer
 * @param InputDelegator LDfield input delegator.
 */
export function initFactory<
  Props extends { [key: string]: string | undefined; },
  ExtraData = never
>(
  InputDelegator: new () => InputDelegator<
    JSX.Element,
    Props,
    ExtraData,
    Constraints<Props>
  >,
) {
  const { GetComponent } = new ReactInputDelegator(InputDelegator);
  // TODO [Future]: Add constraint type parameters
  return function init([props, constraints, data]: InitParams<Props, ExtraData>):
    RendererState<Props, ExtraData> {
    return {
      GetComponent,
      ...GetComponent(props, constraints, data),
      cache: [props],
      props,
    };
  };
}
