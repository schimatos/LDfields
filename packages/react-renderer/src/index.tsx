import React, { useState, useReducer } from 'react';
import type { InputDelegator } from '@ldfields/delegator';
import { InputDelegatorFactory } from '@ldfields/delegator';
import type { FieldProps } from '@ldfields/field-base/types';
import type {
  RendererFactoryProps, ReducerFunction, RendererActions, RendererState, InitParams,
} from '../types';
import { initFactory } from './init';

function reducer<
  Props extends { [key: string]: string | undefined; },
  ExtraData
>(s: RendererState<Props, ExtraData>, a: RendererActions<Props, ExtraData>):
  RendererState<Props, ExtraData> {
  switch (a.type) {
    case 'propUpdate': {
      return { ...s, cache: [a.props, ...s.cache], props: a.props };
    }
    case 'delegate': {
      const delegation = s.GetComponent(a.props, a.constraints, a.data);
      if (!delegation.delegationChange) {
        return { ...s, ...delegation };
      }
      const props: Partial<Props> = {};
      for (const prop in a.props) {
        if (delegation.modifiable[prop]) {
          props[prop] = a.props[prop];
        }
      }
      for (const prop of delegation.required) {
        if (props[prop] === undefined) {
          const value = s.cache.find((cache) => cache[prop])?.[prop];
          if (value !== undefined) {
            props[prop] = value;
          } else {
            // TODO: GET DEFAULT VALUE
          }
        }
      }
      // TODO: Create function with create constraints
      if (a.constraints?.restrictions) {
        for (const p in a.constraints.restrictions) {
          if (
            typeof p === 'string'
            && a.constraints.restrictions[p]?.in !== undefined
            && !a.constraints.restrictions?.[p]?.in.includes(props[p])
          ) {
            props[p as keyof Props] = a.constraints?.restrictions[p]?.in[0];
          }
        }
      }
      return {
        ...s,
        ...delegation,
        props,
      };
    }
    default: {
      const action: never = a;
      throw new Error(`Invalid reducer action ${action}`);
    }
  }
}

function useRendererReducer<
  Props extends { [key: string]: string | undefined; },
  ExtraData = never
>(
  Delegator: new () => InputDelegator<JSX.Element, Props, ExtraData, Record<string, any>>,
  ...props: InitParams<Props, ExtraData>
) {
  return useReducer<
        ReducerFunction<Props, ExtraData>,
        InitParams<Props, ExtraData>
      >(
        reducer,
        props,
        initFactory<Props, ExtraData>(Delegator),
      );
}

function Component(x: any) {
  const [s, d] = useState('');
  return <>{s}</>
}

export function LDfieldRendererFactory<
  Props extends { [key: string]: string | undefined; },
  ExtraData = never
>({
  settings, fields, genericFields, defaultProps,
}: RendererFactoryProps<Props, ExtraData>) {
  const InputDelegator = InputDelegatorFactory(settings, fields, genericFields);

  return function Renderer({
    props, constraints, onChange, data,
  }: FieldProps<Props, ExtraData>) {
    // const [{ Component, ...state }, dispatch] = useRendererReducer<Props, ExtraData>(
    //   InputDelegator, props, constraints, data,
    // );

    function dispatch(x: any) {
      return x;
    }

    const state = {
      props, constraints, data,
    };

    // TODO: add test for error case
    // encountered where this overrides change
    // during render
    // useEffect(() => {
    //   dispatch({ props, constraints, type: 'delegate' });
    // }, [JSON.stringify(props), JSON.stringify(constraints ?? '')]);

    return (
      <fieldset onBlur={() => { onChange(state.props); }}>
        <Component
          props={{ ...defaultProps, ...state.props }}
          data={data}
          onChange={(p: Partial<Props>) => {
            const update: RendererActions<Props, ExtraData> = {
              type: 'propUpdate',
              props: p,
            };
            dispatch(update);
          }}
        />
      </fieldset>
    );
  };
}

// {
//   // @ts-ignore
//   components.delegation.map((Field, i) => {
//     const newProps = {};
//     for (const elem of Field.modifies) {
//       if (typeof props[elem] !== 'string') {
//         newProps[elem] = '';
//       } else {
//         newProps[elem] = props[elem];
//       }
//     }
//     return (
//       <>
//         <Field.Field
//           // TODO: Check key selection
//           // Getting erros with just field.id
//           // which SHOULD NOT HAPPEN
//           key={
//             // shortid()
//             `${Field.id}-${i}-${shortid()}`
//           }
//           props={newProps}
//           constraints={constraints}
//           data={data}
//           onChange={(p) => {
//             const values = { ...props, ...p };
//             let nextComponents = delegator.delegate(
//               values,
//               constraints,
//             );
//             const output = {};
//             for (const elem in values) {
//               // console.log(components);
//               // @ts-ignore
//               if (nextComponents.modifiable[elem]) {
//                 output[elem] = values[elem];
//               }
//             }
//             let changeMade = true;
//             while (changeMade) {
//               changeMade = false;
//               // @ts-ignore
//               nextComponents = delegator.delegate(output, constraints);
//               for (const modifier of nextComponents.required) {
//                 if (
//                   typeof output[modifier] !== 'string'
//                   && typeof propsCache[modifier] === 'string'
//                 ) {
//                   changeMade = true;
//                   output[modifier] = state.cache[modifier];
//                 }
//               }
//             }

//             // TODO: Add 'inferencing' here

//             // @ts-ignore
// dispatch({
//   type: 'propUpdate',
//   props: p,
// });
//             onChange(output);
//           }}
//         />
//       </>
//     );
//   })
// }
