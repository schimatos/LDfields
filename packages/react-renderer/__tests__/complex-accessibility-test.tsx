/**
 * @jest-environment jsdom
 */
import React from 'react';
import { render } from '@testing-library/react';
import { axe, toHaveNoViolations } from 'jest-axe';
import { defaultSettings } from '@ldfields/default-settings';
import renderer from 'react-test-renderer';
import { LDfieldRendererFactory } from '../src';

const DefaultSettingRenderer = LDfieldRendererFactory<{ [key: string]: string }>({
  settings: defaultSettings,
  fields: [{
    priority: 0,
    modifies: ['value'],
    fieldTargets: ['value'],
    supports() {
      return true;
    },
    Field() {
      return <input aria-label="value" />;
    },
  }, {
    priority: 0,
    modifies: ['termType'],
    fieldTargets: ['termType'],
    supports() {
      return true;
    },
    Field() {
      return <input aria-label="termType" />;
    },
  }, {
    priority: 0,
    modifies: ['label'],
    fieldTargets: ['label'],
    supports() {
      return true;
    },
    Field() {
      return <input aria-label="label" />;
    },
  }, {
    priority: 0,
    modifies: ['datatype'],
    fieldTargets: ['datatype'],
    supports() {
      return true;
    },
    Field() {
      return <input aria-label="datatype" />;
    },
  }],
  genericFields: [],
  defaultProps: {},
});

expect.extend(toHaveNoViolations);

describe('Testing the accessibility of fields', () => {
  beforeEach(() => {
    const { getComputedStyle } = window;
    window.getComputedStyle = (elt) => getComputedStyle(elt);
  });

  it('Should have no violations on the default render for empty props', async () => {
    const { container } = render(
      <DefaultSettingRenderer
        onChange={() => {}}
        constraints={{}}
        props={{}}
      />,
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});

describe('Testing the complex field contains an input fields', () => {
  beforeEach(() => {
    const { getComputedStyle } = window;
    window.getComputedStyle = (elt) => getComputedStyle(elt);
  });

  it('Should have no violations on the default render for empty props', async () => {
    const { container } = render(
      <DefaultSettingRenderer
        onChange={() => {}}
        constraints={{}}
        props={{}}
      />,
    );
    expect(container.innerHTML.toString()).toContain('input');
  });
});

describe('Testing the complex field does not change', () => {
  beforeEach(() => {
    const { getComputedStyle } = window;
    window.getComputedStyle = (elt) => getComputedStyle(elt);
  });

  it('Should have no violations on the default render for empty props', async () => {
    const tree = renderer.create(
      <DefaultSettingRenderer
        onChange={() => {}}
        constraints={{}}
        props={{}}
      />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
