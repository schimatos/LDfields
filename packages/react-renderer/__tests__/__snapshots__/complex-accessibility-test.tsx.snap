// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the complex field does not change Should have no violations on the default render for empty props 1`] = `
<fieldset
  onBlur={[Function]}
>
  <input
    aria-label="value"
  />
  <input
    aria-label="termType"
  />
</fieldset>
`;
